numpy/core/src/multiarray/scalartypes.c.src

```diff
@@ -373,29 +374,24 @@ gentype_nonzero_number(PyObject *m1)
 static PyObject *
 gentype_str(PyObject *self)
 {
-    PyObject *arr, *ret = NULL;
-
-    arr = PyArray_FromScalar(self, NULL);
-    if (arr != NULL) {
-        ret = PyObject_Str((PyObject *)arr);
-        Py_DECREF(arr);
+    static PyObject *fallback = NULL;
+    npy_cache_import("numpy.core.arrayprint", "_scalar_fallback", &fallback);
+    if (fallback == NULL) {
+        return NULL;
     }
-    return ret;
+    return PyObject_CallFunction(fallback, "O", self);
```

Comment:
#8963 it doesn't protect against this kind of recursion (I noted that there).

If we disallow this kind formatter, though, it will be a backwards-compatibility break, which I was trying to avoid.

This formatter makes sense if you assume a clear division between array reprs and scalar reprs. If you assume that scalar reprs are totally distinct from arrays, and x is a scalar, then there can't be any recursion problem. 

The problem only arises because we happened to implement a small number of scalar reprs by calling promoting to 0d and calling the array repr, which can be interpreted as an implementation detail.